clusterDeployer.py

```diff
@@ -99,64 +101,61 @@ def _all_hosts_with_workers(self) -> set[ClusterHost]:
-        names = [x.name for x in self._cc.all_vms()]
-        ips = [x.ip for x in self._cc.all_vms()]
-        for e in q[-1][0][1:]:
-            if e.attrib["name"] in names or e.attrib["ip"] in ips:
-                mac = e.attrib["mac"]
-                name = e.attrib["name"]
-                ip = e.attrib["ip"]
-                pre = "virsh net-update default delete ip-dhcp-host"
-                cmd = f"{pre} \"<host mac='{mac}' name='{name}' ip='{ip}'/>\" --live --config"
-                logger.info(self._local_host.hostconn.run(cmd))
-                removed_macs.append(mac)
-
-        fn = "/var/lib/libvirt/dnsmasq/virbr0.status"
-        with open(fn) as f:
-            contents = f.read()
-
-        if contents:
-            j = json.loads(contents)
-            names = [x.name for x in self._cc.all_vms()]
-            logger.info(f'Cleaning up {fn}')
-            logger.info(f'removing hosts with mac in {removed_macs} or name in {names}')
-            filtered = []
-            for entry in j:
-                if entry["mac-address"] in removed_macs:
-                    logger.info(f'Removed host with mac {entry["mac-address"]}')
-                    continue
-                if "hostname" in entry and entry["hostname"] in names:
-                    logger.info(f'Removed host with name {entry["hostname"]}')
-                    continue
-                logger.info(f'Kept entry {entry}')
-                filtered.append(entry)
-
-            logger.info(self._local_host.hostconn.run("virsh net-destroy default"))
-            with open(fn, "w") as f:
-                f.write(json.dumps(filtered, indent=4))
-            logger.info(self._local_host.hostconn.run("virsh net-start default"))
-            logger.info(self._local_host.hostconn.run("systemctl restart libvirtd"))
+        node_configs: list[NodeConfig] = [node.config for h in self._all_hosts_with_masters() for node in h.k8s_master_nodes]
+        self._local_host.bridge.remove_dhcp_entries(node_configs)
 
-        for h in self._all_hosts:
+        for h in self._all_hosts_with_masters():
             h.ensure_not_linked_to_network()
 
         if os.path.exists(self._cc.kubeconfig):
             os.remove(self._cc.kubeconfig)
 
+    def teardown_workers(self) -> None:
+        cluster_name = self._cc.name
+
+        # If workers not in steps (and masters set), teardown the workers to avoid dangling vms.
+        if "workers" in self.steps and "masters" not in self.steps:
+            logger.info(f"Tearing down (some) workers on {cluster_name}")
+        elif "masters" in self.steps:
+            logger.info(f"Tearing down (some) workers on {cluster_name} before tearing down masters")
+        else:
+            return
+
+        for h in self._all_hosts_with_workers():
+            h.teardown_nodes(h.k8s_worker_nodes)
+
+        node_configs: list[NodeConfig] = [node.config for h in self._all_hosts_with_workers() for node in h.k8s_worker_nodes]
```

Comment:
nested for loop in 1 statement. We might want to break this apart, as it is difficult to read.